### PR TITLE
Avoid unnecessary big for-loop when reporting ticker stats stored in GetContext

### PR DESCRIPTION
**Pull request by miasantreble**
Currently in `Version::Get` when reporting ticker stats stored in `GetContext`, there is a big for-loop through all `Ticker` which adds unnecessary cost to overall CPU usage. We can optimize by storing only ticker values that are used in `Get()` calls in a new struct `GetContextStats` since only a small fraction of all tickers are used in `Get()` calls. For comparison, with the new approach we only need to visit 17 values while old approach will require visiting 100+ `Ticker`

### CODE REVIEW

**Comment by ajkr** [comment]
Can you include the benchmark command and amount of CPU savings in the PR description?

File: db/version_set.cc [hunk 1]
```diff
@@ -1000,9 +1000,9 @@ void Version::Get(const ReadOptions& read_options, const LookupKey& k,
     // report the counters before returning
     if (get_context.State() != GetContext::kNotFound &&
         get_context.State() != GetContext::kMerge) {
-      for (uint32_t t = 0; t < Tickers::TICKER_ENUM_MAX; t++) {
-        if (get_context.tickers_value[t] > 0) {
-          RecordTick(db_statistics_, t, get_context.tickers_value[t]);
+      for (auto ticker_pair : get_context.tickers_pairs) {
```
Comment (siying) [inline db/version_set.cc]: Wrap this with a `if (db_statistics_ != nullptr)`

File: db/version_set.cc [hunk 2]
```diff
@@ -1038,9 +1038,9 @@ void Version::Get(const ReadOptions& read_options, const LookupKey& k,
     f = fp.GetNextFile();
   }
 
-  for (uint32_t t = 0; t < Tickers::TICKER_ENUM_MAX; t++) {
-    if (get_context.tickers_value[t] > 0) {
-      RecordTick(db_statistics_, t, get_context.tickers_value[t]);
+  for (auto ticker_pair : get_context.tickers_pairs) {
```
Comment (siying) [inline db/version_set.cc]: Same here. No need to do this for loop if `db_statistics_` is null.

File: table/get_context.h [hunk 3]
```diff
@@ -27,7 +27,7 @@ class GetContext {
     kMerge,  // saver contains the current merge result (the operands)
     kBlobIndex,
   };
-  uint64_t tickers_value[Tickers::TICKER_ENUM_MAX] = {0};
+  autovector<std::pair<Tickers, uint64_t>> tickers_pairs;
```
Comment (siying) [inline table/get_context.h]: Can you figure out how many entries we need and set it up in autovector?

File: table/get_context.cc [hunk 4]
```diff
@@ -88,10 +89,33 @@ void GetContext::SaveValue(const Slice& value, SequenceNumber seq) {
...
+      it->second += static_cast<uint64_t>(val);
+      return;
+    }
   }
-  tickers_value[ticker] += static_cast<uint64_t>(val);
+  ticker_pairs_.push_back(std::make_pair(ticker, static_cast<uint64_t>(val)));
+}
+
+void GetContext::InitTickers() {
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_HIT, 0));
```
Comment (ajkr) [inline table/get_context.cc]: This looks kind of confusing and hard to maintain. Does it give any benefit?

File: table/get_context.cc [hunk 5]
```diff
@@ -88,10 +89,33 @@ void GetContext::SaveValue(const Slice& value, SequenceNumber seq) {
 }
 
 void GetContext::RecordCounters(Tickers ticker, size_t val) {
-  if (ticker == Tickers::TICKER_ENUM_MAX) {
-    return;
+  for (auto it = ticker_pairs_.begin(); it != ticker_pairs_.end(); ++it) {
```
Comment (ajkr) [inline table/get_context.cc]: I don't have intuition why this PR is faster overall -- before we had an array lookup here, and now it's a for-loop. Have you measured it? Make sure to use a fairly large DB as then a single Get() will have to execute this for-loop many times.

File: table/get_context.cc [hunk 6]
```diff
@@ -88,10 +89,33 @@ void GetContext::SaveValue(const Slice& value, SequenceNumber seq) {
...
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_BYTES_READ, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_MISS, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_ADD, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_BYTES_WRITE, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_INDEX_ADD, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_INDEX_BYTES_INSERT, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_DATA_ADD, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_DATA_BYTES_INSERT, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_FILTER_ADD, 0));
+  ticker_pairs_.push_back(std::make_pair<Tickers, uint64_t>(BLOCK_CACHE_FILTER_BYTES_INSERT, 0));
```
Comment (siying) [inline table/get_context.cc]: Wow, I didn't realize the list is that long. How about we make a GetStats object and make them all variables? We already have things like this: https://github.com/facebook/rocksdb/blob/5.11.fb/db/compaction_iteration_stats.h and https://github.com/facebook/rocksdb/blob/5.11.fb/db/internal_stats.h#L133-L161 which don't feel very hard to maintain.

**Comment by miasantreble** [comment]
| Block Size        | # of Entries           | Children  | Self | Symbol |
|---|---|---|---|---|
| 1M(master) | 1M |      0.58%    | 0.43%  | rocksdb::GetContext::SaveValue |
|  |  |      0.46%     | 0.46% | rocksdb::GetContext::RecordCounters |
|  |  |     0.14%      | 0.14%  | rocksdb::GetContext::GetContext |
| 1M(branch) | 1M |      2.30%    | 2.28%    | rocksdb::GetContext::RecordCounters |
| | |    0.78%    | 0.08%  | rocksdb::GetContext::GetContext |
| | |     0.70%    | 0.21% | rocksdb::GetContext::InitTickers |
| | |     0.35%  |   0.22% | rocksdb::GetContext::SaveValue |
| 1M(master) | 10M |  0.55% |  0.42%  | rocksdb::GetContext::SaveValue |
| | |     0.41%  |    0.41%  | rocksdb::GetContext::RecordCounters |
| | |     0.07%   |  0.07% | rocksdb::GetContext::GetContext |
| 1M(master) | 10M | 2.09%  | 2.07%  | rocksdb::GetContext::RecordCounters |
| | |  0.98%  |   0.10%  | rocksdb::GetContext::GetContext |
| | | 0.88%  |   0.28%  | rocksdb::GetContext::InitTickers |
| | |   0.33%  |   0.21% | rocksdb::GetContext::SaveValue |

**Comment by ajkr** [comment]
Is the correct conclusion from the numbers that ReportCounters (this branch) is faster than RecordCounters (master) in the case where the database is small; otherwise it's slower?

**Comment by miasantreble** [comment]
| Block Size        | # of Entries           | Children  | Self | Symbol |
|---|---|---|---|---|
| 1M(master) | 1M |    |   |  |
|||     0.72%   |  0.27%  |   rocksdb::GetContext::ReportCounters_Loop |
|||     0.34%   |  0.26% |   rocksdb::GetContext::SaveValue |
|||     0.12%    | 0.12%  |   rocksdb::GetContext::RecordCounters |
|||     0.11%    | 0.10%  |   rocksdb::GetContext::GetContext |
| 1M(branch) | 1M | |   |  |
|||     0.35%    | 0.26% |  rocksdb::GetContext::SaveValue |
|||     0.35%    | 0.03% |   rocksdb::GetContext::ReportCounters |
|||     0.04%    | 0.04% |   rocksdb::GetContext::GetContext |

db_bench command used: `TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=fillrandom -num=1000000 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -write_buffer_size=1048576 -compression_type=none && TEST_TMPDIR=/dev/shm perf record -g ./db_bench -benchmarks=readrandom -use_existing_db -readonly -statistics -num=1000000 -threads=32 -cache_size=1048576`

`ReportCounters_Loop` is added in master which is a wrapper for the big for-loop:

~~~~
GetContext::ReportCounters_Loop() {
  for (uint32_t t = 0; t < Tickers::TICKER_ENUM_MAX; t++) {
    if (tickers_value[t] > 0) {
      RecordTick(statistics_, t, tickers_value[t]);
    }
  }
}
~~~~
So looks like the new approach is faster for 1M keys + 1M block size case. Tried to run 5M keys + 1M block size but the perf.data becomes too large to open (7GB)

**Comment by siying** [comment]
The approach is good but where is autovector?

**Comment by siying** [comment]
Can you improve your title and summary? It's not clear what "counters" you mean.

**Pull request closed without merging**